zebra/zebra_mlag.c

```diff
@@ -88,8 +650,544 @@ void zebra_mlag_init(void)
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug("%s: Entering..", __func__);
+
+	rc = zebra_mlag_lib_decode_mlag_hdr(s, &mlag_msg);
+	if (rc)
+		return rc;
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug("%s: Decoded msg length:%d..", __func__,
+			   mlag_msg.data_len);
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug("%s: Mlag ProtoBuf encoding of message:%s", __func__,
+			   zebra_mlag_lib_msgid_to_str(mlag_msg.msg_type, buf,
+						       sizeof(buf)));
+	*msg_type = mlag_msg.msg_type;
+	switch (mlag_msg.msg_type) {
+	case MLAG_MROUTE_ADD: {
+		struct mlag_mroute_add msg;
+		ZebraMlagMrouteAdd pay_load = ZEBRA_MLAG_MROUTE_ADD__INIT;
+		uint32_t vrf_name_len = 0;
+
+		rc = zebra_mlag_lib_decode_mroute_add(s, &msg);
+		if (rc)
+			return rc;
+		vrf_name_len = strlen(msg.vrf_name) + 1;
+		pay_load.vrf_name = XMALLOC(MTYPE_MLAG_PBUF, vrf_name_len);
+		strlcpy(pay_load.vrf_name, msg.vrf_name, vrf_name_len);
+		pay_load.source_ip = msg.source_ip;
+		pay_load.group_ip = msg.group_ip;
+		pay_load.cost_to_rp = msg.cost_to_rp;
+		pay_load.owner_id = msg.owner_id;
+		pay_load.am_i_dr = msg.am_i_dr;
+		pay_load.am_i_dual_active = msg.am_i_dual_active;
+		pay_load.vrf_id = msg.vrf_id;
+
+		if (msg.owner_id == MLAG_OWNER_INTERFACE) {
+			vrf_name_len = strlen(msg.intf_name) + 1;
+			pay_load.intf_name =
+				XMALLOC(MTYPE_MLAG_PBUF, vrf_name_len);
+			strlcpy(pay_load.intf_name, msg.intf_name,
+				vrf_name_len);
+		}
+
+		len = zebra_mlag_mroute_add__pack(&pay_load, tmp_buf);
+		XFREE(MTYPE_MLAG_PBUF, pay_load.vrf_name);
+		if (msg.owner_id == MLAG_OWNER_INTERFACE)
+			XFREE(MTYPE_MLAG_PBUF, pay_load.intf_name);
+	} break;
+	case MLAG_MROUTE_DEL: {
+		struct mlag_mroute_del msg;
+		ZebraMlagMrouteDel pay_load = ZEBRA_MLAG_MROUTE_DEL__INIT;
+		uint32_t vrf_name_len = 0;
+
+		rc = zebra_mlag_lib_decode_mroute_del(s, &msg);
+		if (rc)
+			return rc;
+		vrf_name_len = strlen(msg.vrf_name) + 1;
+		pay_load.vrf_name = XMALLOC(MTYPE_MLAG_PBUF, vrf_name_len);
+		strlcpy(pay_load.vrf_name, msg.vrf_name, vrf_name_len);
+		pay_load.source_ip = msg.source_ip;
+		pay_load.group_ip = msg.group_ip;
+		pay_load.owner_id = msg.owner_id;
+		pay_load.vrf_id = msg.vrf_id;
+
+		if (msg.owner_id == MLAG_OWNER_INTERFACE) {
+			vrf_name_len = strlen(msg.intf_name) + 1;
+			pay_load.intf_name =
+				XMALLOC(MTYPE_MLAG_PBUF, vrf_name_len);
+			strlcpy(pay_load.intf_name, msg.intf_name,
+				vrf_name_len);
+		}
+
+		len = zebra_mlag_mroute_del__pack(&pay_load, tmp_buf);
+		XFREE(MTYPE_MLAG_PBUF, pay_load.vrf_name);
+		if (msg.owner_id == MLAG_OWNER_INTERFACE)
+			XFREE(MTYPE_MLAG_PBUF, pay_load.intf_name);
+	} break;
+	case MLAG_MROUTE_ADD_BULK: {
+		struct mlag_mroute_add msg;
+		ZebraMlagMrouteAddBulk Bulk_msg =
+			ZEBRA_MLAG_MROUTE_ADD_BULK__INIT;
+		ZebraMlagMrouteAdd **pay_load = NULL;
+		int i = 0;
+		bool cleanup = false;
+
+		Bulk_msg.n_mroute_add = mlag_msg.msg_cnt;
+		pay_load = XMALLOC(MTYPE_MLAG_PBUF,
+				   sizeof(ZebraMlagMrouteAdd *)
+					   * Bulk_msg.n_mroute_add);
```

Comment:
actually both are same. see Line Number 781, where we are setting "n_mroute_add" to mlag_msg.msg_cnt.

we need this because protobuf knows only about Bulk_msg structure.

Reply:
as a safe check used "mlag_msg.msg_cnt" in Both BULK_ADD & BULK_DEL allocation